File: online/LeetCodeCN/1~100/70.py
# 爬楼梯 https://leetcode-cn.com/problems/climbing-stairs/
# DP

class Solution:
    def __init__(self):
        self.dp = {1: 1, 2: 2}

    def climbStairs(self, n: int) -> int:
        if n in self.dp:
            return self.dp[n]

        self.dp[n] = self.climbStairs(n-1) + self.climbStairs(n-2)
        return self.dp[n]

File: online/LeetCodeCN/1~100/70.java
// 爬楼梯 https://leetcode-cn.com/problems/climbing-stairs/
// DP

class Solution {
    HashMap<Integer, Integer> dp;

    public Solution() {
        dp = new HashMap<Integer, Integer>();
        dp.put(1, 1);
        dp.put(2, 2);
    }

    public int climbStairs(int n) {
        if (dp.containsKey(n)) {
            return dp.get(n);
        }

        int result = climbStairs(n-1) + climbStairs(n-2);
        dp.put(n, result);
        return result;
    }
}
